src/rgb/contract/assignments.rs

```diff
@@ -708,3 +708,813 @@ mod strict_encoding {
+
+        // Extract balanced confidential output amounts
+        let outputs: Vec<Commitment> = balanced
+            .all_state_pedersen()
+            .iter()
+            .map(|confidential| confidential.commitment)
+            .collect();
+
+        // Create confidential input amounts
+        let inputs: Vec<Commitment> = input_revealed
+            .iter()
+            .map(|revealed| revealed.conceal().commitment)
+            .collect();
+
+        // Check sum verification and return the result
+        amount::Confidential::verify_commit_sum(inputs, outputs)
+    }
+
+    #[test]
+    fn test_zero_balance() {
+        let mut rng = thread_rng();
+
+        // test equal inputs and outputs
+        // 4 values equally distributed among ours and theirs allocations
+        // * One zero value case
+        // * All zero value case
+        // * One u64::MAX value case
+        // * All u64::MAX value case
+        // * One zero value + One u64::MAX value case
+        let amounts = vec![
+            [1u64, 2u64, 3u64, u64::MAX],
+            [10u64, 20u64, 30u64, 40u64],
+            [0u64, 197642u64, u64::MAX, 476543u64],
+            [0u64, 0u64, 0u64, 0u64],
+            [u64::MAX, u64::MAX, u64::MAX, u64::MAX],
+        ];
+
+        for vec in amounts.iter() {
+            assert!(compute_zero_balance(vec, vec, 2));
+        }
+
+        // Test when ours is empty
+        assert!(compute_zero_balance(&amounts[1], &amounts[1], 0));
+
+        // Test when theirs is empty
+        assert!(compute_zero_balance(
+            &amounts[1],
+            &amounts[1],
+            amounts[1].len()
+        ));
+
+        // Test random inputs and outputs
+        // Randomly distributed between ours and theirs allocation
+        for _ in 0..10 {
+            // Randomly generate number of amounts between 1 to 20
+            let length = rng.gen_range(1, 20);
+
+            // Randomly fill the amount vector
+            let mut amounts = vec![0; length];
+            for _ in 0..length {
+                // keep the amount value low for faster testing
+                amounts.push(rng.gen_range::<u64>(0, 10));
+            }
+
+            // Create an output amount vector such that
+            // input.sum() = output.sum(), but
+            // input.count() != output.count()
+            // We do this by generating an empty output vector
+            // of random size. Then randomly add 1 to random positions
+            // until the sum of the vector matches with the input sum
+            // This also creates 0 value allocations probabilistically
+
+            let input_sum: u64 = amounts.iter().sum();
+            let mut output_amounts = vec![0u64; rng.gen_range(1, 20)];
+            let output_length = output_amounts.len();
+            for _ in 0..input_sum {
+                output_amounts[rng.gen_range(0, output_length)] += 1u64;
+            }
+
```

Comment:
```suggestion
// Check non-equivalent amounts
assert_eq!(compute_zero_balance(&amounts[..(length-1)], &output_amounts[..(output_length-1)], rng.gen_range(0, output_length), false);
```

Reply:
I didn't get this one. What are we checking here?

Reply:
We are checking non-equivalent amounts: two randomly generated vecs

Reply:
Since we adjusting the last member of both amount arrays to them will have the same sum, if we will use sub-arrays without the last member, these arrays with extremely large probability will be non-equal in their sum